File: protocol/src/test/javascript/zfoojs/packet/ObjectB.js

const ObjectB = function() {
    this.flag = false; // boolean
    this.innerCompatibleValue = 0; // number
};

ObjectB.PROTOCOL_ID = 103;

ObjectB.prototype.protocolId = function() {
    return ObjectB.PROTOCOL_ID;
};

ObjectB.write = function(buffer, packet) {
    if (packet === null) {
        buffer.writeInt(0);
        return;
    }
    const beforeWriteIndex = buffer.getWriteOffset();
    buffer.writeInt(4);
    buffer.writeBoolean(packet.flag);
    buffer.writeInt(packet.innerCompatibleValue);
    buffer.adjustPadding(4, beforeWriteIndex);
};

ObjectB.read = function(buffer) {
    const length = buffer.readInt();
    if (length === 0) {
        return null;
    }
    const beforeReadIndex = buffer.getReadOffset();
    const packet = new ObjectB();
    const result0 = buffer.readBoolean(); 
    packet.flag = result0;
    if (buffer.compatibleRead(beforeReadIndex, length)) {
        const result1 = buffer.readInt();
        packet.innerCompatibleValue = result1;
    }
    if (length > 0) {
        buffer.setReadOffset(beforeReadIndex + length);
    }
    return packet;
};

export default ObjectB;

File: protocol/src/test/rust/zfoorust/objectB.rs
#![allow(dead_code)]
#![allow(unused_imports)]
#![allow(unused_mut)]
#![allow(unused_variables)]
#![allow(non_snake_case)]
#![allow(non_camel_case_types)]
use std::any::Any;
use std::collections::HashMap;
use std::collections::HashSet;
use crate::zfoorust::i_byte_buffer::{IByteBuffer, IPacket};


#[derive(Clone)]
pub struct ObjectB {
    pub flag: bool,
    pub innerCompatibleValue: i32,
}

impl IPacket for ObjectB {
    fn protocolId(&self) -> i16 {
        return 103;
    }
}

impl ObjectB {
    pub fn new() -> ObjectB {
        let packet = ObjectB {
            flag: false,
            innerCompatibleValue: 0,
        };
        return packet;
    }
}

pub fn writeObjectB(buffer: &mut dyn IByteBuffer, packet: &dyn Any) {
    let message = packet.downcast_ref::<ObjectB>().unwrap();
    let beforeWriteIndex = buffer.getWriteOffset();
    buffer.writeInt(4);
    buffer.writeBool(message.flag);
    buffer.writeInt(message.innerCompatibleValue);
    buffer.adjustPadding(4, beforeWriteIndex);
}

pub fn readObjectB(buffer: &mut dyn IByteBuffer) -> Box<dyn Any> {
    let length = buffer.readInt();
    let mut packet = ObjectB::new();
    if length == 0 {
        return Box::new(packet);
    }
    let beforeReadIndex = buffer.getReadOffset();
    let result0 = buffer.readBool(); 
    packet.flag = result0;
    if buffer.compatibleRead(beforeReadIndex, length) {
        let result1 = buffer.readInt();
        packet.innerCompatibleValue = result1;
    }
    if length > 0 {
        buffer.setReadOffset(beforeReadIndex + length);
    }
    return Box::new(packet);
}